runtime/internal/services/reflection.go

```diff
@@ -0,0 +1,101 @@
+package services
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"io"
+
+	queryv1beta1 "cosmossdk.io/api/cosmos/base/query/v1beta1"
+	reflectionv1 "cosmossdk.io/api/cosmos/reflection/v1"
+	"github.com/cosmos/gogoproto/proto"
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	protov2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// ReflectionService implements the cosmos.reflection.v1 service.
+type ReflectionService struct {
+	reflectionv1.UnimplementedReflectionServiceServer
+	files *descriptorpb.FileDescriptorSet
+}
+
+func NewReflectionService() (*ReflectionService, error) {
+	fds := &descriptorpb.FileDescriptorSet{}
+
+	// load gogo proto file descriptors
+	allFds := proto.AllFileDescriptors()
+	haveFileDescriptor := map[string]bool{}
+	for _, compressedBz := range allFds {
+		rdr, err := gzip.NewReader(bytes.NewReader(compressedBz))
+		if err != nil {
+			return nil, err
+		}
+
+		bz, err := io.ReadAll(rdr)
+		if err != nil {
+			return nil, err
+		}
+
+		fd := &descriptorpb.FileDescriptorProto{}
+		err = protov2.Unmarshal(bz, fd)
+		if err != nil {
+			return nil, err
+		}
+
+		fds.File = append(fds.File, fd)
+		haveFileDescriptor[*fd.Name] = true
+	}
+
+	// load any protoregistry file descriptors not in gogo
+	protoregistry.GlobalFiles.RangeFiles(func(fileDescriptor protoreflect.FileDescriptor) bool {
+		if !haveFileDescriptor[fileDescriptor.Path()] {
+			fds.File = append(fds.File, protodesc.ToFileDescriptorProto(fileDescriptor))
+		}
+		return true
+	})
+
+	slices.SortFunc(fds.File, func(x, y *descriptorpb.FileDescriptorProto) bool {
+		return *x.Name < *y.Name
+	})
+
+	return &ReflectionService{files: fds}, nil
+
+}
+
+func (r ReflectionService) FileDescriptors(_ context.Context, request *reflectionv1.FileDescriptorsRequest) (*reflectionv1.FileDescriptorsResponse, error) {
+	pagination := request.Pagination
+	if pagination == nil {
+		return &reflectionv1.FileDescriptorsResponse{File: r.files.File}, nil
+	}
+
+	if len(pagination.Key) > 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "key based pagination is not supported")
+	}
+
+	n := len(r.files.File)
+	offset := int(pagination.Offset)
+	if offset >= n {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid offset")
+	}
+
+	limit := int(pagination.Limit)
+	extent := offset + limit
+	if extent > n {
+		extent = n
+	}
```

Comment:
I feel like it's an overkill to manually handle pagination here. Would it be okay if I took away pagination on this query?

Reply:
@amaurym any thoughts about this?

Reply:
Good point. OK let's remove pagination here. (can always add it back later in a non-breaking way)

Reply:
Making the change in https://github.com/cosmos/cosmos-sdk/pull/13668 so I don't need a go.mod `replace`